sklearn/decomposition/_mf.pyx

```diff
@@ -0,0 +1,286 @@
+
+    indices = np.arange(data.shape[0])
+
+    for current_iter in range(n_iter):
+        random_state.shuffle(indices)
+
+        for value_index in indices:
+            value = data[value_index]
+            i = row[value_index]
+            j = col[value_index]
+
+            estimation = bias_samples[i] + bias_features[j] + np.dot(L[i, :], R[:, j])
+            error = estimation - value
+
+            L_grad = error * R[:, j] + regularization * L[i, :]
+            R_grad = error *  L[i, :] + regularization * R[:, j]
+
+            r_discount = 1
+            l_discount = 1
+
+            if adagrad:
+                feature_grad_norms[j] += np.dot(R_grad, R_grad)
+                sample_grad_norms[i] += np.dot(L_grad, L_grad)
+
+                r_discount = sqrt(feature_grad_norms[j]) + 1e-6
+                l_discount = sqrt(sample_grad_norms[i]) + 1e-6
+
+            R[:, j] -= learning_rate / r_discount * R_grad
+            L[i, :] -= learning_rate / l_discount * L_grad
+
+            if fit_intercept:
+                s_discount = 1
+                f_discount = 1
+
+                if adagrad:
+                    feature_bias_grad_norms[j] += error ** 2
+                    sample_bias_grad_norms[i] += error ** 2
+
+                    s_discount = sqrt(sample_bias_grad_norms[i]) + 1e-6
+                    f_discount = sqrt(feature_bias_grad_norms[j]) + 1e-6
+
+                bias_samples[i] -= learning_rate / s_discount * error
+                bias_features[j] -= learning_rate / f_discount * error
+
+
+        if verbose > 0:
+            _log_step(L, R, bias_features, bias_samples,
+                      row, col, data, n_iter, current_iter,
+                      regularization)
+
+
+cdef rr1(double* bias,
+         np.ndarray[double, ndim=1] w,
+         np.ndarray[double, ndim=1] b,
+         np.ndarray[double, ndim=2] X,
+         np.ndarray[double, ndim=1] y,
+         double lam, int n_iter,
+         int fit_intercept):
+
+    cdef int n_samples = X.shape[0]
+    cdef int n_components = X.shape[1]
+
+    errors = y - np.dot(X, w) - b - bias[0]
+
+    for current_iter in range(n_iter):
+        for k in range(n_components):
+            data = X[:, k]
+            errors += w[k] * data
+            w[k] = 0
+
+            a = np.dot(data, data)
+            d = np.dot(data, errors)
+
+            w[k] = d / (lam + a)
+            errors -= w[k] * data
+
+        if fit_intercept:
+            bias[0] = (y - np.dot(X, w) - b).mean()
+
+
+cdef rr(double* bias,
+         np.ndarray[double, ndim=1] w,
+         np.ndarray[double, ndim=1] b,
+         np.ndarray[double, ndim=2] X,
+         np.ndarray[double, ndim=1] y,
+         double lam,
+         int n_iter,
+         int fit_intercept):
+
+    cdef int n_samples = X.shape[0]
+    cdef int n_components = X.shape[1]
+
+    A = np.linalg.inv(np.dot(X.T, X) + n_samples * lam * np.eye(n_components))
+    D = np.dot(A, X.T)
```

Comment:
There are way more efficient ways to solve a ridge regression than actually inverting the matrix. Actually, `sklearn.linear_model.ridge_regression` implements several, including an iterative conjugate gradient one that could maybe be used in a similar trick (solve the subproblem only partially) for speedups. To me it seems like it would be worth some comparative benchmarks.

Also, this function seems to consist almost entirely of pure python calls, is it even worth cythonizing?